### PR TITLE
Common Pool Deadlock

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
This patch fixes a deadlock situation where the FastObjectLoader
exhausts the common thread pool with parallelStream and then a
invalidateLayout is issued. Since the latter call is scheduled
on the same thread pool, the FastObjectLoader will not progress
because it relies on fetchLayout to complete, but it can't.

Why should this be merged: Fixes a deadlock created by the FastObjectLoader

Related issue(s) (if applicable): #1688 and #1687

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 1]
```diff
@@ -632,8 +643,12 @@ private void findCheckPointsInLogAddress(long address, ILogData logData) {
      *
      */
     private void resurrectCheckpoints() {
-        streamsMetaData.entrySet().parallelStream()
-                .forEach(entry -> {
+        ExecutorService executorService = Executors.newFixedThreadPool(numOfWorkers, new ThreadFactoryBuilder()
```
Comment (StrikingThirteen) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Maybe add a bit comment to the method doc highlighting the behavior that it throws FastObjectException if submit fails.

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Why would submit fail? Tasks are not submitted after shutdown is called

**Comment by Maithem** [comment]
I've update this patch. Please review it again.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 2]
```diff
@@ -632,8 +645,13 @@ private void findCheckPointsInLogAddress(long address, ILogData logData) {
...
+                .setNameFormat("FastObjectLoaderResurrectCheckpointsThread-%d").build());
+        CompletableFuture[] cfs = new CompletableFuture[streamsMetaData.size()];
+        int i = 0;
+        try {
+            for (Map.Entry<UUID, StreamMetaData> entry : streamsMetaData.entrySet()) {
+                cfs[i++] = CompletableFuture.runAsync(() -> {
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Good catch, I dropped it from the last PR.
Addressed.

**Pull request merged**